[Bug]: correctly close response body in rpc

````
api/proxy/client.go:77:23: response body must be closed (bodyclose)
        resp, err := c.Delete(ctx, fmt.Sprintf("%s/cache/erase/%s", host, key))
                             ^
common/rpc/lb.go:170:20: response body must be closed (bodyclose)
        resp, err := c.Get(ctx, url)
                          ^
common/rpc/lb.go:190:19: response body must be closed (bodyclose)
        resp, err := c.Do(ctx, request)
                         ^
common/rpc/lb.go:215:19: response body must be closed (bodyclose)
        resp, err := c.Do(ctx, request)
...
```

--- a/blobstore/api/proxy/client.go
+++ b/blobstore/api/proxy/client.go
@@ -78,5 +78,6 @@ func (c *client) Erase(ctx context.Context, host string, key string) error {
 	if err != nil {
 		return err
 	}
+	defer resp.Body.Close()
 	return rpc.ParseData(resp, nil)
 }

--- a/blobstore/blobnode/core/chunk/compact.go
+++ b/blobstore/blobnode/core/chunk/compact.go
@@ -220,7 +220,6 @@ COMPACT:
 			if err != nil {
 				if err == core.ErrChunkScanEOF {
 					span.Infof("chunk %s scan finished", cs.ID())
-					err = nil
 					break COMPACT
 				} else {
 					span.Errorf("scan chunk meta(%s) failed: %v", cs.ID(), err)

--- a/blobstore/cli/common/cfmt/access_test.go
+++ b/blobstore/cli/common/cfmt/access_test.go
@@ -51,13 +51,13 @@ func TestParseLocation(t *testing.T) {
 		},
 	}
 
-	locx, err := cfmt.ParseLocation("")
+	_, err := cfmt.ParseLocation("")
 	require.Error(t, err)
 
 	_, err = cfmt.ParseLocation("xxxx")
 	require.Error(t, err)
 
-	locx, err = cfmt.ParseLocation("{}")
+	locx, err := cfmt.ParseLocation("{}")
 	require.NoError(t, err)
 	require.Equal(t, access.Location{}, locx)
 

--- a/blobstore/clustermgr/diskmgr/alloc.go
+++ b/blobstore/clustermgr/diskmgr/alloc.go
@@ -260,11 +260,7 @@ func (s *idcStorage) allocFromBlobNodeStorages(ctx context.Context, count int, t
 
 	for _, diskInfo := range excludes {
 		diskInfo.lock.RLock()
-		if excludeHosts == nil {
-			excludeHosts = map[string]bool{diskInfo.info.Host: true}
-		} else {
-			excludeHosts[diskInfo.info.Host] = true
-		}
+		excludeHosts[diskInfo.info.Host] = true
 		diskInfo.lock.RUnlock()
 	}
 

--- a/blobstore/clustermgr/diskmgr/alloc_test.go
+++ b/blobstore/clustermgr/diskmgr/alloc_test.go
@@ -16,7 +16,6 @@ package diskmgr
 
 import (
 	"context"
-	"fmt"
 	"math/rand"
 	"os"
 	"strconv"
@@ -123,7 +122,7 @@ func TestAlloc(t *testing.T) {
 		// refresh cluster's disk space allocator
 		testDiskMgr.refresh(ctx)
 
-		t.Log(fmt.Sprintf("all disk length: %d", len(testDiskMgr.allDisks)))
+		t.Logf("all disk length: %d", len(testDiskMgr.allDisks))
 
 		// alloc from not enough space, alloc should return ErrNoEnoughSpace
 		for _, idc := range testIdcs {
@@ -235,7 +234,7 @@ func TestAllocWithSameHost(t *testing.T) {
 		testDiskMgr.RackAware = false
 		testDiskMgr.refresh(ctx)
 
-		t.Log(fmt.Sprintf("all disk length: %d", len(testDiskMgr.allDisks)))
+		t.Logf("all disk length: %d", len(testDiskMgr.allDisks))
 
 		// alloc from not enough space, alloc should return ErrNoEnoughSpace
 		for _, idc := range testIdcs {

--- a/blobstore/common/rpc/lb.go
+++ b/blobstore/common/rpc/lb.go
@@ -112,11 +112,7 @@ var defaultShouldRetry = func(code int, err error) bool {
 }
 
 func (c *lbClient) Do(ctx context.Context, req *http.Request) (*http.Response, error) {
-	resp, err := c.doCtx(ctx, req)
-	if err != nil {
-		return resp, err
-	}
-	return resp, err
+	return c.doCtx(ctx, req)
 }
 
 func (c *lbClient) Form(ctx context.Context, method, url string, form map[string][]string) (resp *http.Response, err error) {
@@ -162,6 +158,7 @@ func (c *lbClient) DoWith(ctx context.Context, req *http.Request, ret interface{
 	if err != nil {
 		return err
 	}
+	defer resp.Body.Close()
 	err = serverCrcEncodeCheck(ctx, req, resp)
 	if err != nil {
 		return err
@@ -174,7 +171,7 @@ func (c *lbClient) GetWith(ctx context.Context, url string, ret interface{}) err
 	if err != nil {
 		return err
 	}
-	return ParseData(resp, ret)
+	return parseData(resp, ret)
 }
 
 func (c *lbClient) PutWith(ctx context.Context, url string, ret interface{}, params interface{}, opts ...Option) (err error) {
@@ -194,6 +191,7 @@ func (c *lbClient) PutWith(ctx context.Context, url string, ret interface{}, par
 	if err != nil {
 		return
 	}
+	defer resp.Body.Close()
 	err = serverCrcEncodeCheck(ctx, request, resp)
 	if err != nil {
 		return err
@@ -219,6 +217,7 @@ func (c *lbClient) PostWith(ctx context.Context, url string, ret interface{}, pa
 	if err != nil {
 		return err
 	}
+	defer resp.Body.Close()
 
 	// set Header and log errors
 	err = serverCrcEncodeCheck(ctx, request, resp)

--- a/blobstore/common/rpc/lb_test.go
+++ b/blobstore/common/rpc/lb_test.go
@@ -184,7 +184,7 @@ func TestLbClient_Delete(t *testing.T) {
 	result := &ret{}
 	resp, err := client.Delete(ctx, "/get/name?id="+strconv.Itoa(122))
 	require.NoError(t, err)
-	err = ParseData(resp, result)
+	err = parseData(resp, result)
 	require.NoError(t, err)
 	require.NotNil(t, result)
 	client.Close()
@@ -337,7 +337,7 @@ func TestLbClient_Post(t *testing.T) {
 	check, err := client.Post(ctx, "", ret{Name: "test_lb_PostJSON"})
 	require.NoError(t, err)
 	client.Close()
-	err = ParseData(check, result)
+	err = parseData(check, result)
 	require.NoError(t, err)
 	require.NotNil(t, result)
 }

--- a/blobstore/common/rpc/server_test.go
+++ b/blobstore/common/rpc/server_test.go
@@ -462,7 +462,7 @@ func TestServerResponseWith(t *testing.T) {
 		resp := resp(router).ToResponse()
 		require.Equal(t, 204, resp.StatusCode)
 
-		err := ParseData(resp, nil)
+		err := parseData(resp, nil)
 		require.Error(t, err)
 	}
 	{
@@ -474,12 +474,12 @@ func TestServerResponseWith(t *testing.T) {
 		require.Equal(t, 200, resp.StatusCode)
 
 		var r marshalData
-		err := ParseData(resp, &r)
+		err := parseData(resp, &r)
 		require.NoError(t, err)
 		require.Equal(t, "rab oof", r.S)
 
 		resp.Body = io.NopCloser(bytes.NewBuffer(nil))
-		err = ParseData(resp, &r)
+		err = parseData(resp, &r)
 		require.Error(t, err)
 	}
 	{
@@ -491,7 +491,7 @@ func TestServerResponseWith(t *testing.T) {
 		require.Equal(t, 200, resp.StatusCode)
 
 		var r marshalToData
-		err := ParseData(resp, &r)
+		err := parseData(resp, &r)
 		require.NoError(t, err)
 		require.Equal(t, "oof rab", r.S)
 	}
@@ -504,7 +504,7 @@ func TestServerResponseWith(t *testing.T) {
 		require.Equal(t, 200, resp.StatusCode)
 
 		var r marshalToData
-		err := ParseData(resp, &r)
+		err := parseData(resp, &r)
 		require.Error(t, err)
 	}
 }

--- a/blobstore/common/rpc/simple.go
+++ b/blobstore/common/rpc/simple.go
@@ -19,7 +19,6 @@ import (
 	"encoding/json"
 	"fmt"
 	"io"
-	"io/ioutil"
 	"net/http"
 	urllib "net/url"
 	"strings"
@@ -158,6 +157,7 @@ func (c *client) DoWith(ctx context.Context, req *http.Request, ret interface{},
 	if err != nil {
 		return err
 	}
+	defer resp.Body.Close()
 
 	err = serverCrcEncodeCheck(ctx, req, resp)
 	if err != nil {
@@ -171,7 +171,7 @@ func (c *client) GetWith(ctx context.Context, url string, ret interface{}) error
 	if err != nil {
 		return err
 	}
-	return ParseData(resp, ret)
+	return parseData(resp, ret)
 }
 
 func (c *client) PutWith(ctx context.Context, url string, ret interface{}, params interface{}, opts ...Option) (err error) {
@@ -191,6 +191,7 @@ func (c *client) PutWith(ctx context.Context, url string, ret interface{}, param
 	if err != nil {
 		return
 	}
+	defer resp.Body.Close()
 	err = serverCrcEncodeCheck(ctx, request, resp)
 	if err != nil {
 		return err
@@ -216,6 +217,7 @@ func (c *client) PostWith(ctx context.Context, url string, ret interface{}, para
 	if err != nil {
 		return err
 	}
+	defer resp.Body.Close()
 
 	err = serverCrcEncodeCheck(ctx, request, resp)
 	if err != nil {
@@ -292,12 +294,14 @@ func (c *client) doWithCtx(ctx context.Context, req *http.Request) (resp *http.R
 	return
 }
 
-// ParseData parse response with data
+// parseData close response body in this package.
+func parseData(resp *http.Response, data interface{}) (err error) {
+	defer resp.Body.Close()
+	return ParseData(resp, data)
+}
+
+// ParseData parse response with data, close response body by yourself.
 func ParseData(resp *http.Response, data interface{}) (err error) {
-	defer func() {
-		io.Copy(ioutil.Discard, resp.Body)
-		resp.Body.Close()
-	}()
 	if resp.StatusCode/100 == 2 {
 		size := resp.ContentLength
 		if data != nil && size != 0 {

--- a/blobstore/common/rpc/simple_test.go
+++ b/blobstore/common/rpc/simple_test.go
@@ -230,7 +230,7 @@ func TestClient_Post(t *testing.T) {
 	result := &ret{}
 	resp, err := simpleClient.Post(ctx, testServer.URL+"/json", &ret{Name: "TestClient_Post"})
 	require.NoError(t, err)
-	err = ParseData(resp, result)
+	err = parseData(resp, result)
 	require.NoError(t, err)
 	require.NotNil(t, result)
 	require.Equal(t, "TestClient_Post+Test", result.Name)
@@ -268,7 +268,7 @@ func TestClient_Put(t *testing.T) {
 	result := &ret{}
 	resp, err := simpleClient.Put(ctx, testServer.URL+"/json", &ret{Name: "TestClient_Put"})
 	require.NoError(t, err)
-	err = ParseData(resp, result)
+	err = parseData(resp, result)
 	require.NoError(t, err)
 	require.NotNil(t, result)
 	require.Equal(t, "TestClient_Put+Test", result.Name)
@@ -325,7 +325,7 @@ func TestClient_ParseDataWithContentLengthZero(t *testing.T) {
 	require.NoError(t, err)
 	resp.StatusCode = 400
 	resp.ContentLength = 0
-	err = ParseData(resp, result)
+	err = parseData(resp, result)
 	require.Error(t, err)
 }
 
@@ -338,7 +338,7 @@ func TestClient_ParseData(t *testing.T) {
 	require.NoError(t, err)
 	resp.StatusCode = 400
 	resp.ContentLength = 12
-	err = ParseData(resp, result)
+	err = parseData(resp, result)
 	require.Error(t, err)
 }
 
@@ -360,7 +360,7 @@ func TestClient_ResponseClose(t *testing.T) {
 		&ret{Name: "TestClient_ResponseClose"})
 	require.NoError(t, err)
 	resp.Body.Close()
-	err = ParseData(resp, result)
+	err = parseData(resp, result)
 	require.Error(t, err)
 }
 

